src/irmin-pack/inode.ml

```diff
@@ -503,38 +616,53 @@ struct
         | Truncated -> fun target -> Intact target
 
-      let of_hash : type ptr. ptr layout -> hash -> ptr = function
+      let of_key : type ptr. ptr layout -> key -> ptr = function
         | Total -> assert false
-        | Partial _ -> fun hash -> { target = Lazy hash }
-        | Truncated -> fun hash -> Broken hash
+        | Partial _ -> fun key -> { target = Lazy key }
+        | Truncated -> fun key -> Broken (Val_ref.of_key key)
+
+      type ('input, 'output) cps = { f : 'r. 'input -> ('output -> 'r) -> 'r }
+      [@@ocaml.unboxed]
 
       let save :
           type ptr.
-          broken:(hash -> unit) ->
-          save_dirty:(ptr t -> unit) ->
+          broken:(hash, key) cps ->
+          save_dirty:(ptr t, key) cps ->
           clear:bool ->
           ptr layout ->
           ptr ->
           unit =
        fun ~broken ~save_dirty ~clear -> function
-        | Total -> fun (Total_ptr entry) -> (save_dirty [@tailcall]) entry
+        (* Invariant: after returning, we can recover the key from the saved
+           pointer (i.e. [key_exn] does not raise an exception). This is necessary
+           in order to be able to serialise a parent inode (for export) after
+           having saved its children. *)
+        | Total ->
+            fun (Total_ptr entry) ->
+              save_dirty.f entry (fun key ->
+                  Val_ref.promote_exn entry.v_ref key)
         | Partial _ -> (
             function
             | { target = Dirty entry } as box ->
-                if clear then box.target <- Lazy (Lazy.force entry.hash)
-                else
-                  (* Promote from dirty to lazy as it will be saved during
-                     [save_dirty]. *)
-                  box.target <- Lazy_loaded entry;
-                (save_dirty [@tailcall]) entry
+                save_dirty.f entry (fun key ->
+                    if clear then box.target <- Lazy key
+                    else (
+                      box.target <- Lazy_loaded entry;
+                      Val_ref.promote_exn entry.v_ref key))
             | { target = Lazy_loaded entry } as box ->
-                if clear then box.target <- Lazy (Lazy.force entry.hash);
-                (save_dirty [@tailcall]) entry
+                (* TODO(craigfe): document why we save in this case. *)
+                save_dirty.f entry (fun key ->
+                    if clear then box.target <- Lazy key)
             | { target = Lazy _ } -> ())
         | Truncated -> (
             function
-            | Broken h -> (broken [@tailcall]) h
-            | Intact entry -> (save_dirty [@tailcall]) entry)
+            | Intact entry ->
+                save_dirty.f entry (fun key ->
+                    Val_ref.promote_exn entry.v_ref key)
+            | Broken vref ->
+                if not (Val_ref.is_key vref) then
+                  broken.f (Val_ref.to_hash vref) (fun key ->
+                      Val_ref.promote_exn vref key))
```

Comment:
This branch is not covered by the UTs.

Reply:
Good spot. I've added a comment to warn about this; hopefully it'll go away completely when we strip out `Truncated` as you discussed here: https://github.com/mirage/irmin/pull/1659#discussion_r773093699.